Ensure FileStatus and FileSystemEntry Hidden and ReadOnly attributes are retrieved the same way

carlossanlop opened the pull request:
Fixes https://github.com/dotnet/runtime/issues/37301

These two files do very similar tasks, but when it comes to collect the Hidden attribute, the first one is simpler than the second one:

https://github.com/dotnet/runtime/blob/bd6cbe3642f51d70839912a6a666e5de747ad581/src/libraries/System.IO.FileSystem/src/System/IO/Enumeration/FileSystemEntry.Unix.cs#L59-L67

https://github.com/dotnet/runtime/blob/bd6cbe3642f51d70839912a6a666e5de747ad581/src/libraries/System.IO.FileSystem/src/System/IO/FileStatus.Unix.cs#L81-L82

This change ensures both collect the Hidden attribute the same way, and makes it more obvious that both methods are closely related.

Also, the ReadOnly attribute was not being collected in the first file, so I added code to do it, as well as a unit test.

I ran all the System.IO.FileSystem unit tests in Ubuntu and in MacOS and they passed. Pending yet to see the CI results.

This fix needs to go into 5.0, or later it could become a breaking change, as explained [here](https://github.com/dotnet/runtime/issues/37301#issuecomment-649209428).

carlossanlop committed:
Ensure FileStatus and FileSystemEntry IsHidden attribute is retrieved the same way

carlossanlop commented on src/libraries/System.IO.FileSystem/tests/Enumeration/SkipAttributeTests.cs:
```diff
@@ -25,21 +25,24 @@ public void SkippingHiddenFiles()
 
-            // Put a period in front to make it hidden on Unix
-            FileInfo fileTwo = new FileInfo(Path.Combine(testDirectory.FullName, "." + GetTestFileName()));
+            FileInfo fileOne = new FileInfo(Path.Combine(testDirectory.FullName, GetTestFileName()));
```
No changes in this test, just organized some lines for easier understanding.
It also helps to show that this test is affected by my PR.

carlossanlop commented:
@iSazonov can you take a look to this PR?

iSazonov commented on src/libraries/System.IO.FileSystem/src/System/IO/FileStatus.Unix.cs:
```diff
@@ -34,24 +34,32 @@ internal struct FileStatus
+
+        internal static bool IsHidden(Interop.Sys.FileStatus fileStatus)
+        {
+            // If the filename starts with a period or has UF_HIDDEN flag set, it's hidden.
```
It seems the comment should be moved to place where IsHidden() is called.

iSazonov commented on src/libraries/System.IO.FileSystem/tests/Enumeration/SkipAttributeTests.cs:
```diff
@@ -25,21 +25,24 @@ public void SkippingHiddenFiles()
+            FileInfo fileOne = new FileInfo(Path.Combine(testDirectory.FullName, GetTestFileName()));
             FileInfo fileThree = new FileInfo(Path.Combine(testSubdirectory.FullName, GetTestFileName()));
+
+            // Put a period in front of files two and four to make them hidden on Unix
+            FileInfo fileTwo = new FileInfo(Path.Combine(testDirectory.FullName,     "." + GetTestFileName()));
             FileInfo fileFour = new FileInfo(Path.Combine(testSubdirectory.FullName, "." + GetTestFileName()));
```
Are there tests for UserFlags.UF_HIDDEN?

carlossanlop replied:
This is what I am investigating currently.

carlossanlop replied:
I added a unit test to verify this flag in OSX, and I debugged it manually to check that it worked.

iSazonov commented on src/libraries/System.IO.FileSystem/src/System/IO/FileStatus.Unix.cs:
```diff
@@ -65,37 +73,55 @@ internal bool IsReadOnly(ReadOnlySpan<char> path, bool continueOnError = false)
+                (fileName.Length > 0 && fileName[0] == '.') || IsHidden(_fileStatus));
+        }
+
         public void SetAttributes(string path, FileAttributes attributes)
```
As for throw, in line 131 below Hidden attribute presents as valid - I think we could exclude it on systems which do not support this. (I don't know how detect this in right way.)

Jozkee replied:
If `FileAttribute.Hidden` is not supported on Linux, we should update the  docs since they say otherwise https://docs.microsoft.com/dotnet/api/system.io.fileattributes?view=netcore-3.1

carlossanlop replied:
I'm not convinced we have to throw for this.

If the flag is not supported in the current OS, then it will never be set, so any condition checking if the UF_HIDDEN flag was set, will be false.

carlossanlop replied:
See line 147, we are already checking that. No need to throw.

```cs
            if (Interop.Sys.CanSetHiddenFlag)
            {
```

Jozkee commented on src/libraries/System.IO.FileSystem/tests/Enumeration/AttributeTests.cs:
```diff
@@ -123,6 +123,7 @@ public void IsHiddenAttribute()
             fileOne.Create().Dispose();
             fileTwo.Create().Dispose();
+
             if (PlatformDetection.IsWindows)
                 fileTwo.Attributes = fileTwo.Attributes | FileAttributes.Hidden;
```
Shouldn't this also be performed on other platforms?

carlossanlop replied:
No - On Linux you hide a file by prepending a dot to the filename (see line 122 in this file). On Windows, you need to set the flag.

carlossanlop replied:
You pointed out the documentation says it is supported in Linux, Windows and MacOS. Let me degug this unit test in WIndows and Unix to see if it's truly supported there, and if so, then the if should be removed, and maybe a separate test should be added for Linux and MacOS (because of the "." prefix in the name, which already hides the file).

carlossanlop replied:
So in Linux, if I remove the code that prepends a "." to the filename, and attempt to set FileAttributes.Hidden to fileTwo, the test fails, because
`Interop.Sys.CanSetHiddenFlag` returns false.
The documentation needs to be updated for Linux.
I'll test MacOS and Windows now.

carlossanlop replied:
It works on OSX. I actually found a test in FileSystemEntry which would prevent the user from collecting an enumeration of hidden files. I fixed it now.
Next: Windows.

carlossanlop replied:
Works on Windows too.

carlossanlop commented on src/libraries/System.IO.FileSystem/src/System/IO/Enumeration/FileSystemEntry.Unix.cs:
```diff
@@ -46,41 +45,37 @@ namespace System.IO.Enumeration
-            // Same idea as the directory check, just repeated for (and tweaked due to the
-            // nature of) symlinks.
+
+            int resultLStat = Interop.Sys.LStat(entry.FullPath, out Interop.Sys.FileStatus lstatInfo);
```
Not for this change, but we can make the improvement in the future.

carlossanlop committed:
Add missing check in public property FileSystemEntry.IsHidden. Address PR suggestions.

carlossanlop committed:
Add tests for Hidden and ReadOnly attribute check for each platform. Split existing Skip attribute test into different platforms.

carlossanlop commented:
@Jozkee @iSazonov can you take another look? The new commits include:
- Added unit tests for each platform, and split an existing one into platforms: The `FileAttributes.Hidden` flag is only supported by Windows and MacOS. Linux does not support it, the file needs to be hidden using a dot prefix in the filename.
- I found a bug in the `FileSystemEntry.IsHidden` property, where we were only checking for the dot prefix in the filename, but we also had to check for the `IsHidden` attribute.
- Cleaned the code as suggested in the PR.

carlossanlop commented on src/libraries/System.IO.FileSystem/src/System/IO/Enumeration/FileSystemEntry.Unix.cs:
```diff
@@ -143,7 +140,7 @@ public FileAttributes Attributes
         public DateTimeOffset LastWriteTimeUtc => _status.GetLastWriteTime(FullPath, continueOnError: true);
         public bool IsDirectory => _status.InitiallyDirectory;
-        public bool IsHidden => _directoryEntry.Name[0] == '.';
+        public bool IsHidden => _directoryEntry.Name[0] == '.' || (Attributes & FileAttributes.Hidden) != 0;
```
Actually, I think this should be:
```cs
        public bool IsHidden => _directoryEntry.Name[0] == '.' || (Interop.Sys.CanSetHiddenFlag && (Attributes & FileAttributes.Hidden) != 0);
```
Otherwise, if the user sets the Hidden flag in Linux, it won't really hide the file, but this will return true.
I'm going to test this on Linux.

carlossanlop replied:
I added this check too, just in case. I noticed that the Hidden flag cannot be set to the file Attributes in Linux, because the property performs some checks to see if it's possible. But just in case, attempting to read the value should be protected.

carlossanlop replied:
The public property was introduced here: https://github.com/dotnet/runtime/commit/c60340730ce3cb5002001e505e1fb8b360496771

But the change that added the case for UF_HIDDEN should've added this check:
https://github.com/dotnet/corefx/pull/34560

carlossanlop replied:
I had a conversation with @Jozkee about this and the `Interop.Sys.CanSetHiddenFlag` check is not necessary, because the Linux P/Invoke will never allow setting it. So I'm reverting the last commit.

Jozkee replied:
`Attributes` adds `FileAttributes.ReadOnly` which is irrelevant in this operation.
```suggestion
        public bool IsHidden => _directoryEntry.Name[0] == '.' || (_initialAttributes & FileAttributes.Hidden) != 0;
```

Jozkee replied:
Can you please also add a comment on top explaining where the attribute is supported?

carlossanlop replied:
Suggestion applied, and comment added.

Jozkee commented on src/libraries/System.IO.FileSystem/tests/Enumeration/AttributeTests.cs:
```diff
@@ -113,18 +113,42 @@ public void DirectoryAttributesAreExpected()
+        public void IsHiddenAttribute_Windows_OSX()
         {
+            // Put a period in front to make it hidden on Unix
+            IsHiddenAttributeInternal(useDotPrefix: false, useHiddenFlag: true);
```
Can you also add an OSX test that uses the dot prefix?

carlossanlop replied:
The test `IsHiddenAttribute_Unix` in line 127 runs in OSX too (`TestPlatforms.AnyUnix` is `true` in OSX).

iSazonov commented on src/libraries/System.IO.FileSystem/src/System/IO/FileStatus.Unix.cs:
```diff
@@ -39,19 +39,23 @@ internal struct FileStatus
+            return IsReadOnly(_fileStatus);
+        }
+
+        internal static bool IsReadOnly(Interop.Sys.FileStatus fileStatus)
```
Private?

carlossanlop replied:
No, it needs to be accessed by `FileSystemEntry` too, which is in the same assembly.

iSazonov commented on src/libraries/System.IO.FileSystem/src/System/IO/FileStatus.Unix.cs:
```diff
@@ -39,19 +39,23 @@ internal struct FileStatus
         internal bool IsReadOnly(ReadOnlySpan<char> path, bool continueOnError = false)
         {
             EnsureStatInitialized(path, continueOnError);
+            return IsReadOnly(_fileStatus);
```
Why do we re-read if we have Refresh() for this.

carlossanlop replied:
I'm not sure I understand the question.

The functionality of this code has not been changed. I merely moved the contents of the old `IsReadOnly` method into its own internal helper static method that can be accessed by both `FileSystemEntry` and `FileStatus`, to avoid repeating code.

I don't think we should do a `Refresh` call here, that's what `EnsureStatInitialized` is for - it will only perform a `Refresh` if we haven't called it yet.

iSazonov commented on src/libraries/System.IO.FileSystem/src/System/IO/FileStatus.Unix.cs:
```diff
@@ -65,37 +69,57 @@ internal bool IsReadOnly(ReadOnlySpan<char> path, bool continueOnError = false)
+        internal static bool IsDirectory(Interop.Sys.FileStatus fileStatus)
         {
-            // IMPORTANT: Attribute logic must match the logic in FileSystemEntry
+            return (fileStatus.Mode & Interop.Sys.FileTypes.S_IFMT) == Interop.Sys.FileTypes.S_IFDIR;
```
It is better move p/invokes from FileSystemEntry.Initialize() in the place and exclude re-reading too.
The same for all attributes.

carlossanlop replied:
I talked to @Jozkee about this, we can get them moved to a separate helper class, but I would like to do that in a future PR.

carlossanlop committed:
Use _initialAttributes instead of Attributes for IsHidden. Add comment on top.

iSazonov commented:
Thanks for fixing this!

What is Preview we get the fix in?

Jozkee commented:
@iSazonov .NET RC1
